#15 Shift Catalogs (#35)
Inbibo/shift_docs: docs/reference/dev_help/api_standalone.md

                        [Bool,    []]]}
```

>[!NOTE]

> > The *Operators* list from the `catalog` is a list of lists containing two elements. The first one corresponds to the operator class, and the second one is a list of compatible hosts for the operator. For example, if the second element is `["maya", "houdini"]`the operator will be only compatible with those DCC's and work under those contexts. Refer to [Integrations & Resources](../../integration_resources/integrations_resources.md) for more information.

Author: I rewrote the text, let me know if it is clear enough

> > The *Operators* list from the `catalog` is a list of lists containing two elements. The first one corresponds to the operator class, and the second one is a list of compatible hosts for the operator. For example, if the second element is `["maya", "houdini"]`the operator will be only compatible with those DCC's and work under those contexts. Refer to [Integrations & Resources](../../integration_resources/integrations_resources.md) for more information.

Reviewer: Yes, now it's clear. Thanks!

There are two ways of adding a custom catalog to Shift:

- **As a User Catalog**: Use the [Catalog Manager](../catalogs/#the-catalog-manager) to source the catalog file and add it to Shift. This catalog is saved in the user preferences.
- **As a Environment Catalog**: Add the **path to the directory** the containing the catalog file to the `SHIFT_CATALOG_PATH` environment variable. All files in this path containing a `catalog` object will be identified as custom catalogs and loaded by Shift.

## Operator Syntax

All Shift operators inherit from the `SOperator` Shift's Python class. To write a custom operator, the `__init__` and `execute` methods must be overwritten. 

Here is an example of an operator class syntax:

```python
from shift.core.workflow import SOperator
from shift.core.workflow import SPlug
from shift.core.constants import SType
from shift.core.constants import SDirection
 
class MyOperator(SOperator):

> 

Reviewer 2: I would add a docstrings flagging that the docstring of the class will be use like Help test for the operator inside Shift:
```suggestion
    """The docstring provided in the Operator class will be use like the operator help text inside Shift. You can provide a detailed description of the Operator purpose and the requirements for the inputs and outputs here."""
```

Reviewer 2: Hum, thinking on what I would like like an external developer I would like to have it in the example. But If you prefer to just add a note saying that the docstring of the Operator class will became the documentation of the Operator in Shift, It´s ok too.

Author: I added the informaiton in the docstring

                        [Bool,    []]]}
```

>[!NOTE]
> The *Operators* list from the `catalog` is a list of lists containing two elements. The first one corresponds to the operator class, and the second one is a list of compatible hosts for the operator. If no host is specified, the node will be usable in any context. For instance, if the second element is `["maya", "houdini"]` the operator will be only available for those DCC's. Refer to [Integrations & Resources](../../integration_resources/integrations_resources.md) for more information.

There are two ways of adding a custom catalog to Shift:

- **As a User Catalog**: Use the [Catalog Manager](../catalogs/#the-catalog-manager) to source the catalog file and add it to Shift. This catalog is saved in the user preferences.

> - **As a Environment Catalog**: Add the **path to the directory** the containing the catalog file to the `SHIFT_CATALOG_PATH` environment variable. All files in this path containing a `catalog` object will be identified as custom catalogs and loaded by Shift.

Reviewer: ```suggestion
- **As a Environment Catalog**: Add the **path to the directory** containing the catalog file to the `SHIFT_CATALOG_PATH` environment variable. All files in this path containing a `catalog` object will be identified as custom catalogs and loaded by Shift.
```

Author: I changed "As a Environment Catalog" to "As an Environment Catalog" in the suggestion

        super(self.__class__, self).execute(force)
```

> The `SOperator` constructor method takes care of initializing the plug objects and adding them to the operator. Each plug must be initialized specifying a unique code name, its value, type, and a direction which can be either `SDirection.kIn` or `SDirection.kOut`.

Reviewer 3: ```suggestion
The `SOperator` constructor method takes care of initializing the plug objects and adding them to the operator. Each plug must be initialized specifying:
- a unique code name, which will correspond to the name displayed in the UI
- a value, which will correspond to the value the plug is initialized with
- a type, which will define the behaviour of the plug when translating the provided inputs to a Python object
- a direction which can be either `SDirection.kIn` or `SDirection.kOut`.
```

> The **execute** method takes care of defining the list of steps to be performed when the execution of the operator is issued inside Shift. In the example above the execute method picks up the values of the two input plugs and multiplies them together. Finally it stores the result in the output plug.

Reviewer: ```suggestion
The **execute** method takes care of defining the list of steps to be performed when the execution of the operator is issued inside Shift. In the example above, the execute method picks up the values of the two input plugs and multiplies them together. Finally, it stores the result in the output plug.
```

        super(self.__class__, self).execute(force)
```

The `SOperator` constructor method takes care of initializing the plug objects and adding them to the operator. Each plug must be initialized specifying:
- a unique code name, which will correspond to the name displayed in the UI
- a value, which will correspond to the value the plug is initialized with

> - a type, which will define the behaviour of the plug when translating the provided inputs to a Python object

Reviewer 4: Should we maybe list all available types here?

Author: This should be available in the API right?

Reviewer 4: Yes, maybe then we should just add a todo to have a link here to point to that?

Author: At the moment the link can't be added, as the page does not exist and we do not know exactly how doxygen will structure it.
